### PR TITLE
don't call floor for symint unless necessary

### PR DESCRIPTION
**Pull request by ngimel**
Per @ezyang's advice, added magic sym_int method. This works for 1.0 * s0 optimization, but can't evaluate `a>0` for some args, and still misses some optimization that model rewrite achieves, so swin still fails
(rewrite replaces `B = int(windows.shape[0] / (H * W / window_size / window_size))` with `B = (windows.shape[0] // int(H * W / window_size / window_size))` and model passes)

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -412,6 +412,29 @@ def safe_expand(r):
...
+def sym_int_conversion(a):
+    if isinstance(a, sympy.Mul):
+        aa = a.args
+        if len(aa) == 2 and isinstance(aa[0], sympy.Float) and aa[1].is_integer:
+            coef = sympy.Integer(aa[0])
+            print(len(aa), coef)
+            if aa[0] == coef:
+                return (coef * aa[1])
+    # this is wrong, but a > 0 errors out (cannot evaluate truth value)
+    return _sym_floor(a) #if a > 0 else _sym_ceil(a)
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:434]: Yeah, doing this entirely in sympy means you can't guard. I'll push a suggested fix

**Comment by ezyang** [comment]
https://github.com/pytorch/pytorch/pull/94365/commits/ceea7821bdb0c79316b07323fa2ffa32c23b58a9 pushes it out of sympy and into the magic method. PTAL.

**Comment by ngimel** [comment]
Thanks!

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -591,11 +598,32 @@ def unary_magic_impl(self):
...
-            raise
+
+        # Attempt some extra simplification on SymInt
+        if method == "sym_int":
+            out = None
+            if isinstance(expr, sympy.Mul):
+                aa = expr.args
+                if len(aa) == 2 and isinstance(aa[0], sympy.Float) and aa[1].is_integer:
+                    coef = sympy.Integer(aa[0])
+                    if aa[0] == coef:  # structural equality test
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:609]: Hmm, I was wondering why you needed this, but I guess this is to test the scaling factor is not 2.5 and it's a nice integer. Maybe a comment would be nice.

**Review by ezyang — APPROVED** [review]

File: torch/__init__.py [hunk 3]
```diff
@@ -312,6 +312,9 @@ def __sym_max__(self, other):
     def __sym_min__(self, other):
         raise AssertionError("type stub not overridden")
 
+    def __sym_int__(self, other):
```
Comment (ngimel) [inline torch/__init__.py:315]:
```suggestion
    def __sym_int__(self):
```

**Comment by ZainRizvi** [comment]
@pytorchbot revert -m "This looks like it caused some inductor test to start failing: https://hud.pytorch.org/pytorch/pytorch/commit/8a9ea44985725e57cb82f0d978fafae31577ae6d" -c nosignal

Adding the `ciflow\inductor` label to the next version of this change will help validate the issue is properly fixed

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -478,6 +478,11 @@ def safe_expand(r):
     'floordiv': lambda a, b: FloorDiv(a, b),
 }
 
+
+def error():
+    raise AssertionError("shouldn't be hit")
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:483]: dead code now

**Pull request closed without merging**